File: intrinsic/skills/internal/execute_context_impl.py
# Copyright 2023 Intrinsic Innovation LLC

"""ExecuteContext implementation provided by the skill service."""

from typing import Mapping

from intrinsic.motion_planning import motion_planner_client
from intrinsic.resources.proto import resource_handle_pb2
from intrinsic.skills.python import execute_context
from intrinsic.skills.python import skill_canceller
from intrinsic.skills.python import skill_logging_context
from intrinsic.world.python import object_world_client


class ExecuteContextImpl(execute_context.ExecuteContext):
  """ExecuteContext implementation provided by the skill service.

  Attributes:
    canceller: Supports cooperative cancellation of the skill.
    logging_context: The logging context of the execution.
    motion_planner: A client for the motion planning service.
    object_world: A client for interacting with the object world.
    resource_handles: A map of resource names to handles.
  """

  @property
  def canceller(self) -> skill_canceller.SkillCanceller:
    return self._canceller

  @property
  def logging_context(self) -> skill_logging_context.SkillLoggingContext:
    return self._logging_context

  @property
  def motion_planner(self) -> motion_planner_client.MotionPlannerClient:
    return self._motion_planner

  @property
  def object_world(self) -> object_world_client.ObjectWorldClient:
    return self._object_world

  @property
  def resource_handles(
      self,
  ) -> Mapping[str, resource_handle_pb2.ResourceHandle]:
    return self._resource_handles

  def __init__(
      self,
      canceller: skill_canceller.SkillCanceller,
      logging_context: skill_logging_context.SkillLoggingContext,
      motion_planner: motion_planner_client.MotionPlannerClient,
      object_world: object_world_client.ObjectWorldClient,
      resource_handles: dict[str, resource_handle_pb2.ResourceHandle],
  ):
    self._canceller = canceller
    self._logging_context = logging_context
    self._motion_planner = motion_planner
    self._object_world = object_world
    self._resource_handles = resource_handles

File: intrinsic/skills/internal/execute_context_impl.cc
// Copyright 2023 Intrinsic Innovation LLC
// Intrinsic Proprietary and Confidential
// Provided subject to written agreement between the parties.

#include "intrinsic/skills/internal/execute_context_impl.h"

#include <algorithm>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "absl/time/time.h"
#include "grpcpp/grpcpp.h"
#include "intrinsic/icon/release/status_helpers.h"
#include "intrinsic/logging/proto/context.pb.h"
#include "intrinsic/motion_planning/motion_planner_client.h"
#include "intrinsic/motion_planning/proto/motion_planner_service.grpc.pb.h"
#include "intrinsic/skills/cc/equipment_pack.h"
#include "intrinsic/skills/cc/skill_canceller.h"
#include "intrinsic/skills/internal/skill_registry_client_interface.h"
#include "intrinsic/skills/proto/equipment.pb.h"
#include "intrinsic/skills/proto/footprint.pb.h"
#include "intrinsic/skills/proto/skill_service.pb.h"
#include "intrinsic/skills/proto/skills.pb.h"
#include "intrinsic/world/objects/object_world_client.h"
#include "intrinsic/world/proto/object_world_service.grpc.pb.h"

namespace intrinsic {
namespace skills {

using ::intrinsic::world::ObjectWorldClient;

ExecuteContextImpl::ExecuteContextImpl(
    const intrinsic_proto::skills::ExecuteRequest& request,
    std::shared_ptr<ObjectWorldService::StubInterface> object_world_service,
    std::shared_ptr<MotionPlannerService::StubInterface> motion_planner_service,
    EquipmentPack equipment,
    SkillRegistryClientInterface& skill_registry_client,
    std::shared_ptr<SkillCanceller> skill_canceller)
    : world_id_(request.world_id()),
      object_world_service_(std::move(object_world_service)),
      motion_planner_service_(std::move(motion_planner_service)),
      equipment_(std::move(equipment)),
      skill_registry_client_(skill_registry_client),
      skill_canceller_(skill_canceller),
      log_context_(request.context()) {}

absl::StatusOr<ObjectWorldClient> ExecuteContextImpl::GetObjectWorld() {
  return ObjectWorldClient(world_id_, object_world_service_);
}

absl::StatusOr<motion_planning::MotionPlannerClient>
ExecuteContextImpl::GetMotionPlanner() {
  return motion_planning::MotionPlannerClient(world_id_,
                                              motion_planner_service_);
}

const intrinsic_proto::data_logger::Context& ExecuteContextImpl::GetLogContext()
    const {
  return log_context_;
}

}  // namespace skills
}  // namespace intrinsic
